### PR TITLE
new per_user_key signatures

### PR DESCRIPTION
**Pull request by maxtaco**
- renamed from shared_dh_key
- have signing_kid and encryption_kid
- also therefore need a reverse_sig

### CODE REVIEW

File: src/per_user_key.iced [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  sibkid_slot : () -> "signing_kid"
+  _type : () -> constants.sig_types.per_user_key
+  _type_v2 : () -> constants.sig_types_v2.per_user_key
+
+  _v_include_pgp_details : () -> false
+  _required_sections : () -> super().concat(["per_user_key"])
+
+  _find_fields : ({json}) ->
+    if typeof (v = json?.body?.per_user_key?.generation) isnt 'number'
+      new Error "Need per_user_key.generation to be an integer"
```
Comment (oconnor663) [inline src/per_user_key.iced]: Not new code, but do we care to check that it's a positive, whole number?

File: src/per_user_key.iced [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    else if not json?.body?.per_user_key?.encryption_kid?
+      new Error "need an encryption kid"
+    else null
+
+  _v_check : ({json}, cb) ->
+    esc = make_esc cb, "_v_check"
+    err = @_find_fields { json }
+    if not err?
+      await KeyManager.import_public { hex : json.body.per_user_key.signing_kid }, esc defer()
+      await EncKeyManager.import_public { hex : json.body.per_user_key.encryption_kid }, esc defer()
```
Comment (oconnor663) [inline src/per_user_key.iced]: Are these imports just checking that the type prefixes on these KIDs are correct?

File: test/files/util.iced [hunk 3]
```diff
@@ -0,0 +1,18 @@
+
+{prng} = require 'crypto'
+
+exports.new_uid = new_uid = () -> prng(16).toString('hex')
```
Comment (oconnor663) [inline test/files/util.iced]: Is it ok that the UID doesn't come from a hash of the username? Will we ever try to verify that property on these UIDs?

File: test/files/util.iced [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+exports.new_username = new_username = () -> "u_" + prng(5).toString('hex')
+
+exports.skip = 1
+
+exports.new_arg = new_arg = ({km}) ->
```
Comment (oconnor663) [inline test/files/util.iced]: Should we call this `new_sig_arg` or something?

File: test/files/util.iced (hunk 3 above)
Comment (maxtaco) [inline test/files/util.iced]: not checked here! just test code

File: src/per_user_key.iced (hunk 2 above)
Comment (maxtaco) [inline src/per_user_key.iced]: yes!

File: src/per_user_key.iced (hunk 1 above)
Comment (maxtaco) [inline src/per_user_key.iced]: ok!

**Comment by maxtaco** [comment]
thanks for the review

**Pull request merged**